### PR TITLE
[Textual inversion] Relax loading textual inversion

### PR DESCRIPTION
**Pull request by patrickvonplaten**
# What does this PR do?

This PR makes it easier to use textual inversion embeddings with SDXL.

For example:

```py
import hf_image_uploader as hiu
from safetensors.torch import load_file
from huggingface_hub import snapshot_download

from diffusers import DiffusionPipeline
import torch
import json
import os

model_id = "fofr/sdxl-2004"

folder = snapshot_download(model_id)

with open(os.path.join(folder, "special_params.json"), 'r') as json_file:
    data = json.load(json_file)
    token, token_2 = data["TOK"][:4], data["TOK"][4:]

state_dict = load_file(os.path.join(folder, "embeddings.pti"))

text_encoder_sd = state_dict["text_encoders_0"]
text_encoder_2_sd = state_dict["text_encoders_1"]

pipe = DiffusionPipeline.from_pretrained("stabilityai/stable-diffusion-xl-base-1.0", torch_dtype=torch.float16)

pipe.load_textual_inversion(text_encoder_sd, token=token, text_encoder=pipe.text_encoder, tokenizer=pipe.tokenizer)
pipe.load_textual_inversion(text_encoder_2_sd, token=token_2, text_encoder=pipe.text_encoder_2, tokenizer=pipe.tokenizer_2)
pipe.load_lora_weights(os.path.join(folder, "lora.safetensors"))
pipe.to(torch_dtype=torch.float16)
pipe.to("cuda")

prompt = "A photo of a cyberpunk in a living room from 2004 in the style of <s0><s1>"

image = pipe(prompt, cross_attention_kwargs={"scale": 0.8}, num_inference_steps=20).images[0]
```

### CODE REVIEW

**Comment by patrickvonplaten** [comment]
I really don't feel comfortable adding too much logit that's super unique to https://huggingface.co/fofr/sdxl-2004/tree/main to core diffusers. For example:
- textual inv embeddings are stored in one file (they might very well be stored in two files for other implementations)
- the naming of the textual inv embedding file is super weird - the ".pti" ending is not automatically recognized by our safetensors loading
- Inside the textual inversion dict the keys are named "text_encoders_0" and "text_encoders_1". Very specific naming that doesn't really follow any existing convention
- We additionally have a .json file that includes the name of the tokens in a very weird form "<s0><s1>" - making it very hard to extract

Long story short, I would really prefer to not have to create a new complex loading logic for SDXL that is tailored for this very unique checkpointing format.

The above codesnippet works just fine with this PR and can avoid having to install an additional library.

File: src/diffusers/loaders.py [hunk 1]
```diff
@@ -684,6 +686,11 @@ def load_textual_inversion(
             token (`str` or `List[str]`, *optional*):
                 Override the token to use for the textual inversion weights. If `pretrained_model_name_or_path` is a
                 list, then `token` must also be a list of equal length.
+            text_encoder ([`~transformers.CLIPTextModel`], *optional*):
+                Frozen text-encoder ([clip-vit-large-patch14](https://huggingface.co/openai/clip-vit-large-patch14)).
+                If not specified, function will take self.tokenizer.
+            tokenizer ([`~transformers.CLIPTokenizer`], *optional*):
+                A `CLIPTokenizer` to tokenize text. If not specified, function will take self.tokenizer.
```
Comment (sayakpaul) [inline src/diffusers/loaders.py:693]: So, only single text encoder and tokenizer for now? I'd imagine you'd need to touch both the text encoders to squeeze out performance from SDXL. But, I could very well be wrong.

**Comment by sayakpaul** [comment]
@patrickvonplaten you can remove the `import hf_image_uploader as hiu` from the snippet.

File: src/diffusers/loaders.py (hunk 1 above)
Comment (sayakpaul) [reply src/diffusers/loaders.py:693]: Ah nevermind. It was made clear in the snippet.

File: src/diffusers/loaders.py [hunk 2]
```diff
@@ -897,18 +907,18 @@ def load_textual_inversion(
             else:
                 token = loaded_token
 
-            embedding = embedding.to(dtype=self.text_encoder.dtype, device=self.text_encoder.device)
+            embedding = embedding.to(dtype=text_encoder.dtype, device=text_encoder.device)
```
Comment (sayakpaul) [inline src/diffusers/loaders.py:910]: Is it not backward-breaking? For example, will the following work?

```python
from diffusers import StableDiffusionPipeline
import torch

model_id = "runwayml/stable-diffusion-v1-5"
pipe = StableDiffusionPipeline.from_pretrained(model_id, torch_dtype=torch.float16).to("cuda")

pipe.load_textual_inversion("sd-concepts-library/cat-toy")

prompt = "A <cat-toy> backpack"

image = pipe(prompt, num_inference_steps=50).images[0]
image.save("cat-backpack.png")
```

File: src/diffusers/loaders.py [hunk 3]
```diff
@@ -897,18 +907,18 @@ def load_textual_inversion(
...
 
             # 3. Make sure we don't mess up the tokenizer or text encoder
-            vocab = self.tokenizer.get_vocab()
```
Comment (sayakpaul) [inline src/diffusers/loaders.py:903]: Same as above.

File: src/diffusers/loaders.py [hunk 4]
```diff
@@ -897,18 +907,18 @@ def load_textual_inversion(
...
+            vocab = tokenizer.get_vocab()
             if token in vocab:
                 raise ValueError(
                     f"Token {token} already in tokenizer vocabulary. Please choose a different token name or remove {token} and embedding from the tokenizer and text encoder."
                 )
             elif f"{token}_1" in vocab:
                 multi_vector_tokens = [token]
                 i = 1
-                while f"{token}_{i}" in self.tokenizer.added_tokens_encoder:
```
Comment (sayakpaul) [inline src/diffusers/loaders.py:911]: Same as above.

**Review by sayakpaul — COMMENTED** [review]
Design looks clean to me 👌 Left a couple of questions. 

Okay for me to not add tests for now.

File: src/diffusers/loaders.py [hunk 5]
```diff
@@ -757,15 +764,18 @@ def load_textual_inversion(
         ```
 
         """
-        if not hasattr(self, "tokenizer") or not isinstance(self.tokenizer, PreTrainedTokenizer):
+        tokenizer = tokenizer or getattr(self, "tokenizer", None)
+        text_encoder = text_encoder or getattr(self, "text_encoder", None)
```
Comment (patrickvonplaten) [inline src/diffusers/loaders.py:768]: this makes it backwards compatible IMO cc @sayakpaul

**Comment by sayakpaul** [comment]
LGTM!

**Comment by linnanwang** [comment]
@patrickvonplaten thanks, but I'm still quite confused how to load text embedding e.g. EasyNegative here: https://civitai.com/models/7808/easynegative. It seems the example provided only works for fofr/sdxl-2004. Thanks.

**Comment by andysingal** [comment]
@patrickvonplaten @sayakpaul looking for the same, on how to use SDXL with Text-inversion... can i also use Compel with it?

**Comment by sayakpaul** [comment]
@patrickvonplaten can you provide an example here?

**Comment by kadirnar** [comment]
> # What does this PR do?
> This PR makes it easier to use textual inversion embeddings with SDXL.
> 
> For example:
> 
> ```python
> import hf_image_uploader as hiu
> from safetensors.torch import load_file
> from huggingface_hub import snapshot_download
> 
> from diffusers import DiffusionPipeline
> import torch
> import json
> import os
> 
> model_id = "fofr/sdxl-2004"
> 
> folder = snapshot_download(model_id)
> 
> with open(os.path.join(folder, "special_params.json"), 'r') as json_file:
>     data = json.load(json_file)
>     token, token_2 = data["TOK"][:4], data["TOK"][4:]
> 
> state_dict = load_file(os.path.join(folder, "embeddings.pti"))
> 
> text_encoder_sd = state_dict["text_encoders_0"]
> text_encoder_2_sd = state_dict["text_encoders_1"]
> 
> pipe = DiffusionPipeline.from_pretrained("stabilityai/stable-diffusion-xl-base-1.0", torch_dtype=torch.float16)
> 
> pipe.load_textual_inversion(text_encoder_sd, token=token, text_encoder=pipe.text_encoder, tokenizer=pipe.tokenizer)
> pipe.load_textual_inversion(text_encoder_2_sd, token=token_2, text_encoder=pipe.text_encoder_2, tokenizer=pipe.tokenizer_2)
> pipe.load_lora_weights(os.path.join(folder, "lora.safetensors"))
> pipe.to(torch_dtype=torch.float16)
> pipe.to("cuda")
> 
> prompt = "A photo of a cyberpunk in a living room from 2004 in the style of <s0><s1>"
> 
> image = pipe(prompt, cross_attention_kwargs={"scale": 0.8}, num_inference_steps=20).images[0]
> ```

AttributeError: 'StableDiffusionXLControlNetInpaintPipeline' object has no attribute 'load_textual_inversion'
